src/entities/map/ui/lib/cluster.ts

```diff
@@ -0,0 +1,207 @@
+    const scaledLatLng = this.minMaxScaling({ lat, lng });
+    return {
+      lat: Math.abs(scaledLatLng.lat - this.mean.lat) / this.std.lat,
+      lng: Math.abs(scaledLatLng.lng - this.mean.lng) / this.std.lng,
+    };
+  }
+  /**
+   * 특정 마커를 마커 버퍼에 추가 합니다.
+   * 이 때 추가된 마커가 이상값인지 판단하여 이상값 버퍼에 추가 합니다.
+   * ! 이상값으로 판단 된 마커는 클러스터 기준으로 이상값일 뿐
+   * ! 전체 데이터 기준으로 이상값이라고 판단 할 수 없습니다.
+   * ! 오히려 전체 데이터 기준으로 봤을 땐 해당 군집에 어울리는 값일 수 있습니다.
+   * ! 이상값을 판단하는 기준은 Z-Score가 3 이상인 경우로 설정 하였습니다. (3시그마 규칙)
+   */
+  addMarker(marker: T) {
+    const { lat, lng } = marker;
+    // 초기 시행 시에는 따로 Z-Score를 계산하지 않습니다.
+    if (this.std.lat === 0 || this.std.lng === 0) {
+      this.markers.push(marker);
+      return;
+    }
+    const zScore = this.getZScore({ lat, lng });
+    /**
+     * 새로운 군집을 형성 하기 위해 버퍼에 추가합니다.
+     * 이 때 새로운 군집을 형성하는 기준은 이상값이 아닌 마커들로 구성된 군집입니다.
+     */
+    if (zScore.lat > 3 || zScore.lng > 3) {
+      this.outliers.push(marker);
+      return;
+    }
+    this.markers.push(marker);
+  }
+  /**
+   * 마커의 중심점을 재조정 합니다.
+   * 이 때 마커의 중심점이 변경 되었다면 true를 반환 합니다.
+   */
+  revalidateCluster() {
+    // 재조정 전 클러스터의 중심점을 캐싱 합니다.
+    const prevCenter = { ...this.center };
+
+    // 클러스터의 마커들을 이용해 새로운 중심점을 계산합니다.
+    this.center = this.markers.reduce(
+      (center, { lat, lng }) => ({
+        lat: center.lat + lat / this.markers.length,
+        lng: center.lng + lng / this.markers.length,
+      }),
+      { lat: 0, lng: 0 },
+    );
+    // 새로운 중심점을 이용해 통계값을 계산합니다.
+    this.updateStatisticValue();
+    return (
+      prevCenter.lat !== this.center.lat || prevCenter.lng !== this.center.lng
+    );
+  }
+}
+
+export const useKMeansClustering = <T extends LatLng>(
+  NumOfCluster: number,
+  markers?: T[],
+) => {
+  const getCurrentBounds = useGetMapCurrentBounds();
+  const bounds = getCurrentBounds();
+  if (!markers) {
+    return [];
+  }
+
+  // K개의 클러스터를 생성합니다.
+  // TODO 휴리스틱한 방식으로 초기값 뽑기
+  const randomIndexMap: Record<number, boolean> = {};
+  const clusters = Array.from({ length: NumOfCluster }, () => {
+    let randomIndex;
+    do {
+      randomIndex = Math.floor(Math.random() * markers.length);
+    } while (randomIndexMap[randomIndex]);
+    {
+      randomIndexMap[randomIndex] = true;
+    }
+    return new Cluster(markers[randomIndex], bounds);
+  });
+
+  let isChanged = true;
+  while (isChanged) {
+    clusters.forEach((cluster) => cluster.clearMarkers());
+
+    markers.forEach((marker) => {
+      const [, closestClusterIndex] = clusters.reduce(
+        ([minDistance, minClusterIndex], cluster, index) => {
+          const distance = cluster.calculateDistance(marker);
+          return distance < minDistance
+            ? [distance, index]
+            : [minDistance, minClusterIndex];
+        },
+        [Infinity, -1],
+      );
+      // 가장 가까운 클러스터에게 마커를 추가합니다.
+      clusters[closestClusterIndex].addMarker(marker);
+    });
+    // 클러스터의 중심점을 재조정합니다.
+    // 이 때 모든 클러스터의 중심점이 재조정 되지 않았다면 반복문을 종료 합니다.
```

Comment:
clearMarkers 실행으로 인해 clusters들의 마커와 이상치는 빈배열로 초기화되고 center 값은 남아있는 상태에서
각 marker는 본인과 가장 가까운 클러스터를 찾아 마커를 추가하는 걸로 이해했습니다.
맞을까요?

closestClusterIndex를 첫 번째 요소에 두는 건 어떨까요?


```suggestion
    markers.forEach((marker) => {
      const [closestClusterIndex] = clusters.reduce(
        ([minDistance, minClusterIndex], cluster, index) => {
          const distance = cluster.calculateDistance(marker);
          return distance < minDistance
            ? [index, distance]
            : [minClusterIndex, minDistance];
        },
        [-1, Infinity],
      );
      // 가장 가까운 클러스터에게 마커를 추가합니다.
      clusters[closestClusterIndex].addMarker(marker);
    });
```

Reply:
이해하신 바가 명확히 맞습니다 

사용되는 값이 인덱스이니 인덱스 값을 앞에 두자는 말씀이신거죠 ?

그렇게 적용 하겠습니다.

Reply:
지금 작업하며 생각해봤는데 자바스크립트의 배열 프로토타입 메소드는 대부분 `value,index` 순이기에 비슷한 느낌으로 통일시키게 원상태로 가는건 어떨까요 ?

Reply:
>사용되는 값이 인덱스이니 인덱스 값을 앞에 두자는 말씀이신거죠 ?

네 맞습니다.

>자바스크립트의 배열 프로토타입 메소드는 대부분 value,index 순이기에 비슷한 느낌으로 통일시키게 원상태로 가는건 어떨까요 ?

저는 저건 메소드의 파라미터랑 다른 경우라고 생각했어요. `[minDistance, minClusterIndex]`은 결과값이지 메소드의 파라미터라고 생각을 안했거든요.

근데 용현님 말씀처럼 reduce의 파라미터 순서를 `[value, index], value, index`로 맞추고 싶으시다면 그 의견도 동의합니다 👍